pkg/controller/identity/iampolicy/controller.go

```diff
@@ -0,0 +1,299 @@
+	if err := c.kube.Get(ctx, n, s); err != nil {
+		return nil, errors.Wrap(err, errGetProviderSecret)
+	}
+
+	policyClient, err := c.newClientFn(ctx, s.Data[p.Spec.CredentialsSecretRef.Key], p.Spec.Region, awsclients.UseProviderSecret)
+	return &external{client: policyClient, kube: c.kube}, errors.Wrap(err, errCreatePolicyClient)
+}
+
+type external struct {
+	client iam.PolicyClient
+	kube   client.Client
+}
+
+func (e *external) Observe(ctx context.Context, mgd resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mgd.(*v1alpha1.IAMPolicy)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errUnexpectedObject)
+	}
+
+	if cr.Status.AtProvider.Arn == "" {
+		return managed.ExternalObservation{}, nil
+	}
+
+	policyResp, err := e.client.GetPolicyRequest(&awsiam.GetPolicyInput{
+		PolicyArn: aws.String(cr.Status.AtProvider.Arn),
+	}).Send(ctx)
+
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(resource.Ignore(iam.IsErrorNotFound, err), errGet)
+	}
+
+	if policyResp.Policy == nil {
+		return managed.ExternalObservation{}, errors.New(errEmptyPolicy)
+	}
+	policy := policyResp.Policy
+
+	cr.SetConditions(runtimev1alpha1.Available())
+
+	cr.Status.AtProvider = v1alpha1.IAMPolicyObservation{
+		Arn:                           aws.StringValue(policy.Arn),
+		AttachmentCount:               aws.Int64Value(policy.AttachmentCount),
+		DefaultVersionID:              aws.StringValue(policy.DefaultVersionId),
+		IsAttachable:                  aws.BoolValue(policy.IsAttachable),
+		PermissionsBoundaryUsageCount: aws.Int64Value(policy.PermissionsBoundaryUsageCount),
+		PolicyID:                      aws.StringValue(policy.PolicyId),
+	}
+
+	versionRsp, err := e.client.GetPolicyVersionRequest(&awsiam.GetPolicyVersionInput{
+		PolicyArn: aws.String(cr.Status.AtProvider.Arn),
+		VersionId: aws.String(cr.Status.AtProvider.DefaultVersionID),
+	}).Send(ctx)
+
+	if err != nil || versionRsp.PolicyVersion == nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, errPolicyVersion)
+	}
+
+	update, err := iam.IsPolicyUpToDate(cr.Spec.ForProvider, *versionRsp.PolicyVersion)
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, errUpToDate)
+	}
+
+	return managed.ExternalObservation{
+		ResourceExists:   true,
+		ResourceUpToDate: update,
+	}, nil
+}
+
+func (e *external) Create(ctx context.Context, mgd resource.Managed) (managed.ExternalCreation, error) {
+	cr, ok := mgd.(*v1alpha1.IAMPolicy)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errUnexpectedObject)
+	}
+
+	cr.Status.SetConditions(runtimev1alpha1.Creating())
+
+	createResp, err := e.client.CreatePolicyRequest(&awsiam.CreatePolicyInput{
+		Description:    cr.Spec.ForProvider.Description,
+		Path:           cr.Spec.ForProvider.Path,
+		PolicyDocument: aws.String(cr.Spec.ForProvider.Document),
+		PolicyName:     aws.String(meta.GetExternalName(cr)),
+	}).Send(ctx)
```

Comment:
This seems like a tricky API in that the create call takes a user-specified name, but returns a fully qualified ARN that must be stored and provided when observing, updating, or deleting the policy. Currently we're storing that ARN in `cr.Status.AtProvider.ARN`, but I think we should store in in the `crossplane.io/external-name` annotation too.

Kubernetes requires us to be able to recover if we lose object status for some reason, so we can't rely on the ARN being stored there. For example if you back up this CR using https://velero.io/ the status will be removed when the CR is restored to the cluster.

```go
meta.SetExternalName(cr, aws.StringValue(createResp.CreatePolicyOutput.Policy.Arn))
if err := e.kube.Update(ctx, cr); err != nil {
   // ... handle error
}

// We must update status _after_ the above Update call, because the Update call overwrites
// cr.Status with whatever is currently in the API server.
cr.Status.AtProvider.Arn = aws.StringValue(createResp.CreatePolicyOutput.Policy.Arn)
return managed.ExternalCreation{}, errors.Wrap(e.kube.Status().Update(ctx, cr), errCreate)
```

I believe we can add something like the above at line 188, then just use `meta.GetExternalName` in all API calls. It will return only the name for the first observe and create, then return the ARN on subsequent observes, updates, and deletes.

Reply:
@negz As per our conversation on Slack https://crossplane.slack.com/archives/CEF5N8X08/p1588019841267300 , I am just concerned about the case where the external-name is reset by a second `kubectl apply`

Reply:
Good point! As discussed on our call earlier it looks like we address a similar case in the SecurityGroup resource by storing its (unqualified) name in `spec`, then storing the unique identifier that the AWS API returns as our external name. In this case the user would be expected to never set the external name annotation.


https://github.com/crossplane/provider-aws/blob/211791/pkg/controller/network/securitygroup/controller.go#L142

Reply:
Thanks @negz I've made the required change.